### PR TITLE
[foreach] bump tensor's version and define backward via torchgen (as possible)

### PR DESCRIPTION
**Pull request by crcrpar**
## summary
- increment tensor versions in inplace foreach functions
- add a logic to take care of `ArrayRef<Scalar>`

rel: https://github.com/pytorch/pytorch/issues/58833, https://github.com/pytorch/pytorch/pull/89591

cc @ptrblck

### CODE REVIEW

**Comment by crcrpar** [comment]
this could fix https://github.com/pytorch/pytorch/pull/91846#issuecomment-1412565392

File: test/test_foreach.py [hunk 1]
```diff
@@ -65,6 +65,25 @@ def __call__(self, inputs, is_cuda, is_fastpath, **kwargs):
...
+            (num_tensors, num_tensors), dtype=dtype, device=device,
+            requires_grad=True, noncontiguous=not is_fastpath,
+        )
+    return transform
+
+def clone(arg):
+    if isinstance(arg, (list, tuple)):
+        return [clone(a) for a in arg]
+    if torch.is_tensor(arg):
+        return arg.clone().detach().requires_grad_()
```
Comment (albanD) [inline test/test_foreach.py:82]: nit: `.requires_grad_(arg.requires_grad)`

**Review by albanD — COMMENTED** [review]
Are there more new formulas than OpInfo updated?
Also the codegen is changing the Node binding accessor right? The accessor test should be updated no?

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -2963,3 +2963,56 @@
 - name: _reshape_copy(Tensor self, SymInt[] size) -> Tensor
   self: grad.reshape_symint(self.sym_sizes())
   result: auto_linear
+
+# note(crcrpar): `torchgen/api/autograd` logic would unwantedly replace substrings of `self` and `other` of function names.
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:2967]: Are all the notes here things we want to fix before merging?

Comment (crcrpar) [reply tools/autograd/derivatives.yaml:2967]: I don't think these are high priority or easy to fix.

this first note: I didn't feel like changing the reference function name
the third (maximum and minimum): would you want to add overloads to the native functions?

Comment (albanD) [reply tools/autograd/derivatives.yaml:2967]: These two sound good to leave as is.
What about the second? weird numerical errors sound bad?

Comment (crcrpar) [reply tools/autograd/derivatives.yaml:2967]: The numerical errors looked to be caused by some naitve mistakes in torchgen changes, e.g. the grads of `_foreach_sub` seemed to be wrongly negated. I however didn't come up with a fix after comparing the generated code of `_foreach_sub` and `sub` in torch/csrc/autograd/generated/Function.cpp`

**Review by albanD — APPROVED** [review]
Ok! Let's merge this as it fixes a lot of things.
We should look into fixing all these weird errors though I think.

**Comment by crcrpar** [comment]
this could be the cause of https://github.com/pytorch/pytorch/issues/95237

**Comment by crcrpar** [comment]
hmm, all the input tensors could store all the tensors involved in forward path for the backward.
I will look into e.g. tools/autograd/gen_variable_type.py

**Pull request closed without merging**